File: cpp/poly.cpp
#include <algorithm>
#include <iterator>
#include <vector>
#include <unordered_set>
#include <iostream>

#include "poly.h"


// TODO Super naive implementation
bool Poly::operator==(const Poly& other) const {
    Poly a(*this), b(other);

    std::sort(a.mMonoms.begin(), a.mMonoms.end());
    std::sort(b.mMonoms.begin(), b.mMonoms.end());
    return a.mMonoms == b.mMonoms;
}

Poly Poly::operator+(const Poly& b) const {
    std::vector<Monom> monoms;
    std::set_symmetric_difference(
        mMonoms.begin(), mMonoms.end(),
        b.mMonoms.begin(), b.mMonoms.end(),
        std::back_inserter(monoms)
    );

    return Poly(monoms);
}

Poly Poly::operator*(const Monom& b) const {
    if (isZero() || b.isZero()) {
        return Poly();
    }
    std::unordered_set<Monom, Monom::hash> monoms_set;

    for (auto&& m: mMonoms) {
        Monom tmp(m*b);

        if (monoms_set.find(tmp) == monoms_set.end()) {
            monoms_set.insert(tmp);
        } else {
            monoms_set.erase(tmp);
        }
    }

    std::vector<Monom> monoms_vec(std::begin(monoms_set), std::end(monoms_set));

    return Poly(monoms_vec);
}

Poly Poly::operator*(const Poly& b) const {
    if (isZero() || b.isZero()) {
        // return 0
        return Poly();
    }
    std::unordered_set<Monom, Monom::hash> monoms_set;
    for (auto&& m1: mMonoms) {
        for (auto&& m2: b.mMonoms) {
            Monom tmp(m1*m2);
            if (monoms_set.find(tmp) == monoms_set.end()) {
                monoms_set.insert(tmp);
            } else {
                monoms_set.erase(tmp);
            }
        }
    }
    std::vector<Monom> monoms_vec(std::begin(monoms_set), std::end(monoms_set));
    return Poly(monoms_vec);
}

std::ostream& operator<<(std::ostream& out, const Poly &a) {
    if (a.isZero()) {
        out << "0";
    } else if (a.isOne()) {
        out << "1";
    } else {
        for (auto&& m: a.mMonoms) {
            out << m << " ";
        }
    }
    return out;
}

Poly spoly(const Poly& f, const Poly& g) {
    Monom f_lm = f.lm();
    Monom g_lm = g.lm();
    Monom lcm = f_lm*g_lm;

    Poly spoly;

    spoly = f*(lcm/f_lm) + g*(lcm/g_lm);

    return spoly;
}

Poly normalform(const Poly& f, const std::vector<Poly>& F) {
    Poly p(f), r;

    if (F.empty()) {
        return r;
    }

    while (!p.isZero()) {
        int i {0};
        bool divisionoccured {false};
        Monom p_lm, fi_lm;
        while (i < F.size() && !divisionoccured) {
            p_lm = p.lm();
            fi_lm = F[i].lm();
            if (p_lm.isdivisible(fi_lm)) {
                p = p + F[i]*(p_lm/fi_lm);
                divisionoccured = true;
            } else {
                i++;
            }
        }
        if (!divisionoccured) {
            r = r + Poly(p_lm);
            p = p + Poly(p_lm);
        }
    }
    return r;
}

File: python/poly.py
"""
Boolean Polynomials
"""

import collections
import operator
import itertools

from monom import Monom


class Poly(list):

    one = None
    zero = None

    def __new__(cls, monoms):
        return super(Poly, cls).__new__(cls, monoms)

    def __init__(self, monoms):
        return super(Poly, self).__init__(sorted(monoms, reverse=True))

    def __add__(self, other):
        # symmetric difference
        if isinstance(other, Monom):
            return Poly(set(self) ^ set([other]))
        elif isinstance(other, Poly):
            return Poly(set(self) ^ set(other))
        else:
            return NotImplemented

    def __mul__(self, other):
        if isinstance(other, Monom):
            if self == Poly.zero or other.isZero():
                return Poly.zero
            monoms = map(lambda m: m*other, self)
        elif isinstance(other, Poly):
            if self == Poly.zero or other == Poly.zero:
                return Poly.zero
            monoms = itertools.starmap(
                operator.mul,
                (itertools.product(self, other))
            )
        else:
            return NotImplemented

        counter = collections.Counter(monoms)
        return Poly({m for m, c in counter.items() if c % 2 != 0})

    def __str__(self):
        if self == Poly.zero:
            return "0"
        return " + ".join(map(str, sorted(self, reverse=True)))

    def lm(self):
        if self == Poly.zero:
            return Monom.zero
        return self[0]

    @staticmethod
    def S(f, g):
        """
        Return s-polynomial
        """
        f_lm, g_lm = f.lm(), g.lm()
        lcm = f_lm.lcm(g_lm)
        spoly = f*(lcm/f_lm) + g*(lcm/g_lm)
        return spoly

    def NF(self, F):
        """
        NormalForm w.r.t F
        D. Cox, J. Little, D. O'Shea - Ideals, Varieties, and Algorithms
        $3 A Division Algorithm
        """
        p = Poly(self)
        r = Poly.zero

        if F == []:
            return p

        while p != Poly.zero:
            i = 0
            divisionoccured = False
            while i < len(F) and (not divisionoccured):
                p_lm, fi_lm = p.lm(), F[i].lm()
                if p_lm.isdivisible(fi_lm):
                    p = p + F[i]*(p_lm/fi_lm)
                    divisionoccured = True
                else:
                    i += 1
            if not divisionoccured:
                r = r + p_lm
                p = p + p_lm

        return r

Poly.zero = Poly([])
Poly.one = Poly([Monom.one])
